### PR TITLE
:bug: Fakeclient: Support WithWatch

### PR DESCRIPTION
**Pull request by alvaroaleman**
When adding client.WithWatch, we forgot to extend the fakeclient to also
support that, this change adds that support.

/assign @vincepri 

WDYT about changing the existing constructors to return a `client.WithWatch` vs adding a new set of constructors? I opted for extending them, because its easier to use, as `client.WithWatch` is a superset of `client.Client`, so will work as a replacement for both.

### CODE REVIEW

**Comment by vincepri** [comment]
Would it makes sense to just have Watch within the Client interface?

**Comment by alvaroaleman** [comment]
> Would it makes sense to just have Watch within the Client interface?

I think that would confuse a lot of ppl and make them try to use that for getting cached data. In the context of a cache-backed client it would also result in establishing a second watch. The usecase for this are mainly CLIs that wait for an event to happen.

**Comment by vincepri** [comment]
Got it, that makes sense. Having the constructors to return a complete interface makes sense to me, although I wonder in that case if can also cause confusion.

**Comment by alvaroaleman** [comment]
> Got it, that makes sense. Having the constructors to return a complete interface makes sense to me, although I wonder in that case if can also cause confusion.

Yeah, that is what I was unsure about as well. Explicitly splitting here means three new constructors though.

File: pkg/client/fake/client_test.go [hunk 1]
```diff
@@ -541,6 +543,32 @@ var _ = Describe("Fake client", func() {
...
+			objWatch, err := cl.Watch(context.Background(), &corev1.ServiceList{})
+			Expect(err).NotTo(HaveOccurred())
+
+			defer objWatch.Stop()
+
+			go func() {
+				defer GinkgoRecover()
+				// It is likely starting a new goroutine is slower than progressing
+				// in the outer routine, sleep to make sure this is always true
+				time.Sleep(100 * time.Millisecond)
```
Comment (stevekuznetsov) [inline pkg/client/fake/client_test.go:557]: Why not a channel or semaphore to enforce it?

Comment (alvaroaleman) [reply pkg/client/fake/client_test.go:557]: What we are waiting for is that the main routine starts to read, how would you sync that with a channel or semaphore?

**Review by vincepri — APPROVED** [review]

**Pull request merged**